File: depend/src/depend.f90
program depend
  use src_file_m
  implicit none

  type(src_file), allocatable :: files(:)
  type(vector_src_item)       :: programs
  type(vector_src_item)       :: modules
  type(vector_src_item)       :: submodules
  type(vector_string)         :: filenames, include_folders
  integer                     :: i, j, k, l, expect
  logical                     :: file_exists
  type(src_item)              :: tmp
  character(len=SLEN)         :: arg, incfolders, folder, filename, buildfolder

  ! process command line arguments
  expect = 0
  call filenames%init(0, c = 16)
  incfolders = ""
  buildfolder = ""
  do i = 1, command_argument_count()
    call get_command_argument(i, arg)

    if (expect > 0) then
      if (expect == 1) then
        buildfolder = trim(arg)
      elseif (expect == 2) then
        incfolders = trim(arg)
      end if
      expect = 0
    else
      if (len_trim(arg) >= 2) then
        if (arg(1:2) == "-b") then
          if (len_trim(arg) > 2) then
            buildfolder = arg(3:len_trim(arg))
          else
            expect = 1
          end if
        elseif (arg(1:2) == "-I") then
          if (len_trim(arg) > 2) then
            incfolders = arg(3:len_trim(arg))
          else
            expect = 2
          end if
        else
          filename = trim(arg)
          call filenames%push(filename)
        end if
      end if
    end if
  end do

  ! include folders, separated by colons
  call include_folders%init(0, c = 8)
  if (incfolders /= "") then
    do while (.true.)
      i = scan(trim(incfolders), ':')
      if (i == 0) then
        folder = trim(incfolders)
        call include_folders%push(folder)
        exit
      else
        if ((i == 1) .or. (i == len_trim(incfolders))) call error("include folder parse error")
        folder = trim(incfolders(1:i-1))
        call include_folders%push(folder)
        incfolders = trim(adjustl(incfolders(i+1:len_trim(incfolders))))
      end if
    end do
  end if

  ! init files
  allocate (files(filenames%n))
  do i = 1, size(files)
    arg = filenames%d(i)

    ! get slash
    j = scan(trim(arg), '/', back = .true.)
    if (j == 0) then
      folder = ""
      filename = arg
    else
      folder = trim(adjustl(arg(1:j)))
      filename = trim(adjustl(arg(j+1:len_trim(arg))))
    end if
    call files(i)%init(folder, filename)
  end do

  ! collect all programs, modules and submodules
  call programs%init(  0, c = 8)
  call modules%init(   0, c = 8)
  call submodules%init(0, c = 8)
  do i = 1, size(files)
    if (files(i)%program_name /= "") then
      do j = 1, programs%n
        if (programs%d(j)%name == files(i)%program_name) then
          call error("program name multiple times defined. name: "//files(i)%program_name)
        end if
      end do
      tmp%name = files(i)%program_name
      tmp%file_index = i
      call programs%push(tmp)
    end if

    do j = 1, files(i)%modules%n
      do k = 1, modules%n
        if (modules%d(k)%name == files(i)%modules%d(j)) then
          call error("module name defined multiple times. name: "//files(i)%modules%d(j))
        end if
      end do
      tmp%name = trim(files(i)%modules%d(j))
      tmp%file_index = i
      call modules%push(tmp)
    end do
    do j = 1, files(i)%submodules%n
      do k = 1, submodules%n
        if (submodules%d(k)%name == files(i)%submodules%d(j)) then
          call error("submodule name defined multiple times. name: "//files(i)%submodules%d(j))
        end if
      end do
      tmp%name = trim(files(i)%submodules%d(j))
      tmp%file_index = i
      call submodules%push(tmp)
    end do
  end do

  ! get dependency indices
  do i = 1, size(files)
    do j = 1, files(i)%use_modules%n
      do k = 1, modules%n
        if (modules%d(k)%name == files(i)%use_modules%d(j)) then
          files(i)%dep_modules(j) = k

          ! add anchor file of this module (if it does not exist already)
          do l = 1, files(i)%dep_anchor%n
            if (files(i)%dep_anchor%d(l) == modules%d(k)%file_index) exit
          end do
          if (l > files(i)%dep_anchor%n) call files(i)%dep_anchor%push(modules%d(k)%file_index)

          exit
        end if
      end do
    end do

    ! search include files in include folders
    do j = 1, files(i)%includes%n
      ! search file folder first
      inquire(file = trim(files(i)%folder)//trim(files(i)%includes%d(j)), exist = file_exists)
      if (file_exists) then
        files(i)%includes%d(j) = trim(files(i)%folder)//trim(files(i)%includes%d(j))
      else
        ! search include folders in order
        do k = 1, include_folders%n
          inquire(file = trim(include_folders%d(k))//trim(files(i)%includes%d(j)), exist = file_exists)
          if (file_exists) then
            files(i)%includes%d(j) = trim(include_folders%d(k))//trim(files(i)%includes%d(j))
            exit
          end if
        end do
      end if
      ! do not include file if not found in folders (e.g. includes of a library interface)
      if (.not. file_exists) then
        files(i)%includes%d(j) = ""
      end if
    end do
  end do

  ! output targets variable (programs)
  write(output_unit, "(1A)", advance = "no") "TARGETS ="
  do i = 1, programs%n
    write(output_unit, "(1A)", advance = "no") " "//trim(buildfolder)//trim(programs%d(i)%name)
  end do
  write(output_unit,*)
  write(output_unit,*)

  ! output object list for each target
  do i = 1, programs%n
    write(output_unit, "(1A)", advance = "no") "OBJECTS_"//trim(programs%d(i)%name)//" ="
    do j = 1, size(files)
      if ((files(j)%program_name == "") .or. (files(j)%program_name == programs%d(i)%name)) then
        write(output_unit, "(1A)", advance = "no") " "//trim(buildfolder)//trim(files(j)%objname)
      end if
    end do
    write(output_unit,*)
  end do
  write(output_unit,*)

  ! output anchor list for each target
  do i = 1, programs%n
    write(output_unit, "(1A)", advance = "no") "ANCHORS_"//trim(programs%d(i)%name)//" ="
    do j = 1, size(files)
      if ((files(j)%program_name == "") .or. (files(j)%program_name == programs%d(i)%name)) then
        write(output_unit, "(1A)", advance = "no") " "//trim(buildfolder)//trim(files(j)%ancname)
      end if
    end do
    write(output_unit,*)
  end do
  write(output_unit,*)

  ! output targets
  do i = 1, programs%n
    write(output_unit, "(1A)") ".PHONY: "//trim(programs%d(i)%name)
    write(output_unit, "(1A)") trim(buildfolder)//trim(programs%d(i)%name)//": $(ANCHORS_"//trim(programs%d(i)%name)//") $(OBJECTS_"//trim(programs%d(i)%name)//") $(OBJECTS_C) $(LIBS)"
    write(output_unit, "(1A)") char(9)//'@printf "%b" "$(FC_COL)$(FC)$(NO_COL) $(FFLAGS) -I'//trim(buildfolder)//' -o $(OU_COL)'//trim(buildfolder)//trim(programs%d(i)%name)//'$(NO_COL) $(IN_COL)$(OBJECTS_'//trim(programs%d(i)%name)//') $(OBJECTS_C) $(LIBS)$(NO_COL)\n\n"'
    write(output_unit, "(1A)") char(9)//"@$(FC) $(FFLAGS) -I"//trim(buildfolder)//" -o "//trim(buildfolder)//trim(programs%d(i)%name)//" $(OBJECTS_"//trim(programs%d(i)%name)//") $(OBJECTS_C) $(LIBS)"
    write(output_unit,*)
  end do

  ! output anchor files
  do i = 1, size(files)
    write(output_unit, "(1A)", advance="no") trim(buildfolder)//trim(files(i)%ancname)//" : "//trim(files(i)%folder)//trim(files(i)%filename)

    ! depend on include files
    do k = 1, files(i)%includes%n
      if (files(i)%includes%d(k) /= "") then
        write(output_unit, "(1A)", advance="no") " "//trim(files(i)%includes%d(k))
      end if
    end do

    ! depend on anchor files
    do k = 1, files(i)%dep_anchor%n
      write(output_unit, "(1A)", advance = "no") " "//trim(buildfolder)//trim(files(files(i)%dep_anchor%d(k))%ancname)
    end do

    ! next line
    write(output_unit,*)
  end do

  ! output object files
  do i = 1, size(files)
    write(output_unit, "(1A)", advance="no") trim(buildfolder)//trim(files(i)%objname)//" : "//trim(files(i)%folder)//trim(files(i)%filename)

    ! depend on include files
    do k = 1, files(i)%includes%n
      if (files(i)%includes%d(k) /= "") then
        write(output_unit, "(1A)", advance="no") " "//trim(files(i)%includes%d(k))
      end if
    end do

    ! depend on anchor files
    do k = 1, files(i)%dep_anchor%n
      write(output_unit, "(1A)", advance = "no") " "//trim(buildfolder)//trim(files(files(i)%dep_anchor%d(k))%ancname)
    end do

    ! next line
    write (output_unit,*)
  end do
end program

File: depend.py
import os, argparse, collections, sys, re

# regex programs
re_include   = re.compile(r"m4_include\(\s*(.*)\)")
re_program   = re.compile(r"^program *(.*?)(?: *!|$)")
re_module    = re.compile(r"^module *(.*?)(?: *!|$)")
re_submodule = re.compile(r"^submodule *\((.*)\) *(.*?)(?: *!|$)")
re_use       = re.compile(r"^use(?:.*::)? *(.*?)(?: *,|$| *!)")
re_library   = re.compile(r"^\!\!\!library *(.*)")

class ParseException(Exception):
  def __init__(self, file, line_number, line, msg):
    self.file        = file
    self.line_number = line_number
    self.line        = line
    self.msg         = msg

class src_item:
  def __init__(self, name, file_index):
    self.name       = name
    self.file_index = file_index

class m4_file:
  def __init__(self, folder, filename):
    self.folder   = folder
    self.filename = filename
    self.includes = []

    file, extension = os.path.splitext(filename)
    if (extension != ".f90"): raise Exception("file extension must be *.f90")

    full_filename = folder + "/" + filename

    # open file
    with open(full_filename) as f:
      # start at the top
      line_number = 0

      # read and parse lines to get includes
      while True:
        line = f.readline()
        if (line == ""): break
        line_number = line_number + 1

        # remove leading and trailing spaces
        line = line.strip(" \r\n")

        # search for m4_include
        result = re_include.match(line)
        if (result is None): continue

        # add included file to includes list
        inc = result.group(1)
        if (inc == ""): raise ParseException(full_filename, line_number, line, "m4_include filename empty")
        if (inc not in self.includes): self.includes.append(inc)

class f90_file:
  def __init__(self, folder, filename):
    self.folder       = folder
    self.filename     = filename
    self.program_name = ""
    self.library_name = ""
    self.modules      = []
    self.submodules   = []
    self.use_modules  = []
    self.dep_anchor   = []
    self.parents      = []

    # object and anchor name
    file, extension = os.path.splitext(filename)
    if (extension != ".f90"): raise Exception("filename must be *.f90")
    self.objname = file + ".o"
    self.ancname = file + ".anc"

    full_filename = folder + "/" + filename

    # open file
    with open(full_filename) as f:
      # start at the top
      line_number = 0

      # read and parse lines
      while True:
        line = f.readline()
        if (line == ""): break
        line_number = line_number + 1

        # remove leading and trailing spaces
        line = line.strip(" \r\n")

        # check for library statement
        result = re_library.match(line)
        if (result is not None):
          if (result.group(1) == ""): raise ParseException(full_filename, line_number, line, "library name missing")
          if (self.library_name != ""): raise ParseException(full_filename, line_number, line, "multiple library statements in one file")
          if (self.program_name != ""): raise ParseException(full_filename, line_number, line, "program and library statement in one file")
          self.library_name = result.group(1).lower()

          continue

        # check for program statement
        result = re_program.match(line)
        if (result is not None):
          if (result.group(1) == ""): raise ParseException(full_filename, line_number, line, "program name missing")
          if (self.program_name != ""): raise ParseException(full_filename, line_number, line, "multiple program statements in one file")
          if (self.library_name != ""): raise ParseException(full_filename, line_number, line, "program and library statement in one file")
          self.program_name = result.group(1).lower()

          continue

        # check for module statement
        result = re_module.match(line)
        if (result is not None):
          if (result.group(1) == ""): raise ParseException(full_filename, line_number, line, "module name missing")
          module_name = result.group(1).lower()

          # make sure this is not a "module subroutine ..." etc. statement
          if (module_name.find(" ") == -1):
            if (module_name not in self.modules): self.modules.append(module_name)

          continue

        # check for submodule statement
        result = re_submodule.match(line)
        if (result is not None):
          if (result.group(1) == ""): raise ParseException(full_filename, line_number, line, "submodule parent missing")
          parent = result.group(1).lower()
          if (result.group(2) == ""): raise ParseException(full_filename, line_number, line, "submodule name missing")
          submodule_name = result.group(2).lower()
          submodule_name = parent + "@" + submodule_name
          if (parent not in self.use_modules): self.use_modules.append(parent)
          if (parent not in self.parents): self.parents.append(parent)
          if (submodule_name not in self.submodules): self.submodules.append(submodule_name)

          continue

        # check for use statement
        result = re_use.match(line)
        if (result is not None):
          if (result.group(1) == ""): raise ParseException(full_filename, line_number, line, "use module name missing")
          module_name = result.group(1).lower()
          if (module_name not in self.use_modules): self.use_modules.append(module_name)

          continue

    # init module dependency index list
    self.dep_modules = [-1] * len(self.use_modules)

# parse command line arguments
parser = argparse.ArgumentParser()
parser.add_argument("filenames", nargs = "+")
args = parser.parse_args()

files = []
for f in args.filenames:
  folder, filename = os.path.split(f)

  # m4
  file = m4_file(folder, filename)
  print("$(BUILD_DIR)/" + filename + ": " + folder + "/" + filename, end = "")
  for i in range(len(file.includes)):
    if (os.path.isfile(folder + "/" + file.includes[i])):
      print (" " + folder + "/" + file.includes[i], end = "")
  print("")

  # collect f90 files
  files.append(f90_file(folder, filename))

# collect all programs, libraries, modules and submodules
programs   = []
libraries  = []
modules    = []
submodules = []
for i in range(len(files)):
  if (files[i].program_name != ""):
    for j in range(len(programs)):
      if (files[i].program_name == programs[j].name): raise Exception("program name " + programs[j].name + " found more than once")
    programs.append(src_item(files[i].program_name, i))

  if (files[i].library_name != ""):
    for j in range(len(libraries)):
      if (files[i].library_name == libraries[j].name): raise Exception("library name " + libraries[j].name + " found more than once")
    libraries.append(src_item(files[i].library_name, i))

  for j in range(len(files[i].modules)):
    for k in range(len(modules)):
      if (modules[k].name == files[i].modules[j]): raise Exception("module name " + modules[k].name + " found more than once")
    modules.append(src_item(files[i].modules[j], i))

  for j in range(len(files[i].submodules)):
    for k in range(len(submodules)):
      if (submodules[k].name == files[i].submodules[j]): raise Exception("submodule name " + submodules[k].name + " found more than once")
    submodules.append(src_item(files[i].submodules[j], i))

# get dependency indices
for i in range(len(files)):
  for j in range(len(files[i].use_modules)):
    for k in range(len(modules)):
      if (modules[k].name == files[i].use_modules[j]):
        files[i].dep_modules[j] = k

        # add anchor file of this module (if it does not exist already)
        if (modules[k].file_index not in files[i].dep_anchor):
          files[i].dep_anchor.append(modules[k].file_index)

        break

# get object list for each program
pobjects = []
for i in range(len(programs)):
  pobjects.append([programs[i].file_index])

  # init deque of file candidates
  file_deque = collections.deque()
  for d in files[programs[i].file_index].dep_anchor:
    file_deque.append(d)

  # add files and their dependencies
  while (len(file_deque) > 0):
    j = file_deque.popleft()

    # add file if not already in list
    if (j not in pobjects[i]): pobjects[i].append(j)

    # add dependencies
    for d in files[j].dep_anchor:
      file_deque.append(d)

  # add submodules
  for s in submodules:
    for j in range(len(files[s.file_index].parents)):
      for k in pobjects[i]:
        if (files[s.file_index].parents[j] in files[k].modules):
          if (s.file_index not in pobjects[i]): pobjects[i].append(s.file_index)

# get object list for each library
lobjects = []
for i in range(len(libraries)):
  lobjects.append([libraries[i].file_index])

  # init deque of file candidates
  file_deque = collections.deque()
  for d in files[libraries[i].file_index].dep_anchor:
    file_deque.append(d)

  # add files and their dependencies
  while (len(file_deque) > 0):
    j = file_deque.popleft()

    # add file if not already in list
    if (j not in lobjects[i]): lobjects[i].append(j)

    # add dependencies
    for d in files[j].dep_anchor:
      file_deque.append(d)

  # add submodules
  for s in submodules:
    for j in range(len(files[s.file_index].parents)):
      for k in lobjects[i]:
        if (files[s.file_index].parents[j] in files[k].modules):
          if (s.file_index not in lobjects[i]): lobjects[i].append(s.file_index)

# output anchor files
print("")
for i in range(len(files)):
  "$(BUILD_DIR)/"
  print("$(BUILD_DIR)/" + files[i].ancname + ": $(BUILD_DIR)/" + files[i].filename, end = "")

  # depend on anchor files
  for j in range(len(files[i].dep_anchor)):
    print(" $(BUILD_DIR)/" + files[files[i].dep_anchor[j]].ancname, end = "")

  # next line
  print("")

# output object files
print("")
for i in range(len(files)):
  print("$(BUILD_DIR)/" + files[i].objname + ": $(BUILD_DIR)/" + files[i].filename, end = "")

  # depend on anchor files
  for j in range(len(files[i].dep_anchor)):
    print(" $(BUILD_DIR)/" + files[files[i].dep_anchor[j]].ancname, end = "")

  # next line
  print("")

# output anchor list for each target
print("")
for i in range(len(programs)):
  print("ANCHORS_" + programs[i].name + " =", end = "")
  for j in pobjects[i]:
    print(" $(BUILD_DIR)/" + files[j].ancname, end = "")
  print("")
for i in range(len(libraries)):
  print("ANCHORS_" + libraries[i].name + " =", end = "")
  for j in lobjects[i]:
    print(" $(BUILD_DIR)/" + files[j].ancname, end = "")
  print("")

# output object list for each target
print("")
for i in range(len(programs)):
  print("OBJECTS_" + programs[i].name + " =", end = "")
  for j in pobjects[i]:
    print(" $(BUILD_DIR)/" + files[j].objname, end = "")
  print("")
for i in range(len(libraries)):
  print("OBJECTS_" + libraries[i].name + " =", end = "")
  for j in lobjects[i]:
    print(" $(BUILD_DIR)/" + files[j].objname, end = "")
  print("")

# output targets variable (programs and libraries)
print("")
print("TARGETS =", end = "")
for i in range(len(programs)):
  print(" $(BUILD_DIR)/" + programs[i].name, end = "")
for i in range(len(libraries)):
  print(" $(BUILD_DIR)/" + libraries[i].name, end = "")

# output targets
print("\n")
for i in range(len(programs)):
  print(".PHONY: " + programs[i].name)
  print("$(BUILD_DIR)/" + programs[i].name + ": $(ANCHORS_" + programs[i].name + ") $(OBJECTS_" + programs[i].name + ") $(C_OBJECTS)")
  print('\t@printf "%b" "$(FC_P) $(FFLAGS) $(FINCLUDE) -o $(OUT_COLOR)' + "$(BUILD_DIR)/" + programs[i].name + '$(OFF_COLOR) $(INP_COLOR)$(OBJECTS_' + programs[i].name + ') $(C_OBJECTS) $(LIBS)$(OFF_COLOR)\\n\\n"')
  print("\t@$(FC) $(FFLAGS) $(FINCLUDE) -o $(BUILD_DIR)/" + programs[i].name + " $(OBJECTS_" + programs[i].name + ") $(C_OBJECTS) $(LIBS)")
  print("")
for i in range(len(libraries)):
  print(".PHONY: " + libraries[i].name)
  print("$(BUILD_DIR)/" + libraries[i].name + ": $(ANCHORS_" + libraries[i].name + ") $(OBJECTS_" + libraries[i].name + ") $(C_OBJECTS)")
  print('\t@printf "%b" "$(AR_P) rcs $(OUT_COLOR)' + "$(BUILD_DIR)/" + libraries[i].name + '.a$(OFF_COLOR) $(INP_COLOR)$(OBJECTS_' + libraries[i].name + ') $(C_OBJECTS)$(OFF_COLOR)\\n\\n"')
  print("\t@ar rcs $(BUILD_DIR)/" + libraries[i].name + ".a $(OBJECTS_" + libraries[i].name + ") $(C_OBJECTS)")
  print("")
